java/java.mx.project/src/org/netbeans/modules/java/mx/project/SuiteSources.java

```diff
@@ -0,0 +1,1195 @@
+            }
+        }
+
+        @Override
+        public void addPropertyChangeListener(PropertyChangeListener pl) {
+            support.addPropertyChangeListener(pl);
+        }
+
+        @Override
+        public void removePropertyChangeListener(PropertyChangeListener pl) {
+            support.removePropertyChangeListener(pl);
+        }
+
+        @Override
+        public SuiteSources owner() {
+            return SuiteSources.this;
+        }
+    }
+
+    private class JdkLibrary extends Library {
+        JdkLibrary(String libName, MxLibrary lib) {
+            super(libName, lib);
+        }
+
+        @Override
+        File getJar(boolean dumpIfMissing) {
+            File first = null;
+            for (File jdk : jdks()) {
+                File jre = new File(jdk, "jre");
+                File jrePath = new File(jre, lib.path().replace('/', File.separatorChar));
+                if (jrePath.exists()) {
+                    return jrePath;
+                }
+
+                if (first == null) {
+                    first = jrePath;
+                }
+
+                File jdkPath = new File(jdk, lib.path().replace('/', File.separatorChar));
+
+                if (jdkPath.exists()) {
+                    return jdkPath;
+                }
+
+            }
+
+            if (dumpIfMissing) {
+                for (File jdk : jdks()) {
+                    File libPath = new File(jdk, lib.path().replace('/', File.separatorChar));
+                    if (!libPath.exists()) {
+                        LOG.log(Level.WARNING, "{0} does not exist", libPath);
+                    } else {
+                        StringBuilder sb = new StringBuilder();
+                        sb.append(libPath).append(" exists:\n");
+                        String[] kids = libPath.list();
+                        if (kids != null) {
+                            for (String n : kids) {
+                                sb.append("  ").append(n).append("\n");
+                            }
+                        }
+                        LOG.log(Level.INFO, sb.toString());
+                        break;
+                    }
+                }
+            }
+            return first;
+        }
+
+        @Override
+        public String getName() {
+            return libName;
+        }
+
+        @Override
+        public Collection<String> depNames() {
+            return lib.dependencies();
+        }
+
+        @Override
+        public Collection<Dep> allDeps() {
+            return allDeps;
+        }
+
+        @Override
+        public void setAllDeps(Collection<Dep> set) {
+            this.allDeps = set;
+        }
+
+        @Override
+        public Set<ClassPath.Flag> getFlags() {
+            return exists ? Collections.emptySet() : Collections.singleton(ClassPath.Flag.INCOMPLETE);
+        }
+
+        @Override
+        public List<? extends PathResourceImplementation> getResources() {
```

Comment:
Can be somehow shared with `Library` implementation ?

Reply:
Excellent idea: af12108